Shareit ТЗ-15 add-bookings

SlavaTurovski opened the pull request.

SlavaTurovski committed:
Выполнение ТЗ-15

SlavaTurovski committed:
Update Comment.java

afomkina commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+        userStorage.findById(userId)
+                .orElseThrow(() -> new NotFoundException("Пользователь не найден: " + userId));
+        List<Booking> userBookings = switch (state) {
+            case CURRENT -> bookingStorage.findCurrentByBookerId(userId, LocalDateTime.now());
```
_Задание со "*"_
Данный блок кода можно оптимизировать с помощью паттерна проектирования [Стратегия](https://refactoring.guru/ru/design-patterns/strategy).
Предлагаю изучить этот паттерн и попробовать реализовать его.

В коммерческой разработке очень часто используются паттерны проектирования, так как код становится более чистым и структурированным. А еще его становится проще поддерживать и масштабировать. 

А еще их часто спрашивают на собеседованиях 😉

afomkina commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        List<Booking> ownerBookings = switch (state) {
+            case CURRENT -> bookingStorage.findCurrentByOwnerId(ownerId, LocalDateTime.now());
```
Здесь тоже отлично подойдет стратегия

afomkina commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,65 +2,148 @@
     }
 
     @Override
     public List<ItemDto> getItemsByOwner(Long userId) {
```
Хорошо, что подтягиваешь комментарии и бронирования батчем. Но выбор last/next сейчас некорректен: в мапы попадает «последняя встретившаяся» запись; нужно считать last как бронь с максимальным end < now (или start < now) и next как с минимальным start > now, фильтруя только APPROVED. Дополнительно: при пустом списке items сразу возвращай List.of()

SlavaTurovski committed:
правки после ревью

- поправил getItemsByOwner в классе ItemServiceImpl
- адаптировал шаблон "стретегия" для booking

SlavaTurovski committed:
Update BookingServiceImpl.java

afomkina commented on src/main/java/ru/practicum/shareit/booking/factory/BookingQueryStrategyFactory.java:
```diff
@@ -0,0 +1,26 @@
+
+public class BookingQueryStrategyFactory {
+
+    public static BookingQueryStrategy create(State state, BookingStorage storage) {
```
Сейчас фабрика фактически просто прячет внутри себя `switch(state)` и вручную создаёт нужную стратегию через new `XxxStrategy(storage)`. Из-за этого:
 - логика выбора всё равно остаётся завязанной на `switch`;
 - в каждой стратегии дублируется одинаковый конструктор с BookingStorage.

Чтобы сделать использование Strategy действительно гибким и избавиться от дублирования, предлагаю:
 -  заменить интерфейс BookingQueryStrategy на абстрактный класс, содержащий общее поле `BookingStorage`, базовый конструктор и метод `State getState()` (каждая стратегия сама «знает», к какому состоянию относится);
 - объявить конкретные стратегии как @Component, чтобы их создавал Spring;
 - в BookingQueryStrategyFactory внедрять все стратегии в виде списка и выбирать нужную по state:

```
BookingQueryStrategy strategy = strategies.stream()
        .filter(s -> s.getState() == state)
        .findFirst()
        .orElseThrow(...);
```

Такой подход полностью убирает фабрику и switch, исключает дублирование конструкторов и делает код расширяемым. Чтобы добавить новую стратегию, нужно только создать новый компонент, не изменяя сервис или фабрику.

SlavaTurovski committed:
правки стратегии

SlavaTurovski committed:
Update DefaultBookingStrategy.java

afomkina reviewed (approved):
Получилось здорово, молодец 👍
